test/test_topology.ml

```diff
@@ -0,0 +1,215 @@
+  let name = "test_topology"
+end)
+
+let make_numa ~numa ~sockets:_ ~cores =
+  let distances =
+    Array.init numa (fun i ->
+        Array.init numa (fun j -> 10 + (11 * abs (j - i))))
+  in
+  let cores_per_numa = cores / numa in
+  let cpu_to_node = Array.init cores (fun core -> core / cores_per_numa) in
+  (cores, NUMA.v ~distances ~cpu_to_node)
+
+let make_numa_amd ~cores_per_numa =
+  (* e.g. AMD Opteron 6272 *)
+  let numa = 8 in
+  let distances =
+    [| [|10; 16; 16; 22; 16; 22; 16; 22|]
+     ; [|16; 10; 22; 16; 16; 22; 22; 16|]
+     ; [|16; 22; 10; 16; 16; 16; 16; 16|]
+     ; [|22; 16; 16; 10; 16; 16; 22; 22|]
+     ; [|16; 16; 16; 16; 10; 16; 16; 22|]
+     ; [|22; 22; 16; 16; 16; 10; 22; 16|]
+     ; [|16; 22; 16; 22; 16; 22; 10; 16|]
+     ; [|22; 16; 16; 22; 22; 16; 16; 10|] |]
+  in
+  let cpu_to_node =
+    Array.init (cores_per_numa * numa) (fun core -> core / cores_per_numa)
+  in
+  (cores_per_numa * numa, NUMA.v ~distances ~cpu_to_node)
+
+type t = {worst: int; average: float; nodes: NUMA.node list; best: int}
+
+let pp =
+  Fmt.(
+    Dump.record
+      [ Dump.field "worst" (fun t -> t.worst) int
+      ; Dump.field "average" (fun t -> t.average) float
+      ; Dump.field "nodes" (fun t -> t.nodes) (Dump.list NUMA.pp_dump_node)
+      ; Dump.field "best" (fun t -> t.best) int ])
+
+let sum_costs l =
+  D.debug "====" ;
+  List.fold_left
+    (fun accum cost ->
+      { worst= max accum.worst cost.worst
+      ; average= accum.average +. cost.average
+      ; nodes= cost.nodes @ accum.nodes
+      ; best= min accum.best cost.best })
+    {worst= min_int; average= 0.; nodes= []; best= max_int}
+    l
+
+let vm_access_costs host all_vms (vcpus, nodes, cpuset) =
+  let nodes = List.of_seq nodes in
+  let all_vms = ((vcpus, nodes), cpuset) :: all_vms in
+  let n = List.length nodes in
+  let costs =
+    cpuset |> CPUSet.elements
+    |> List.map (fun c ->
+           let distances =
+             List.map
+               (fun node -> NUMA.distance host (NUMA.node_of_cpu host c) node)
+               nodes
+           in
+           let worst = List.fold_left max 0 distances in
+           let best = List.fold_left min max_int distances in
+           let average = float (List.fold_left ( + ) 0 distances) /. float n in
+           {worst; best; nodes= []; average})
+    |> sum_costs
+  in
+  D.debug "Costs: %s" (Fmt.to_to_string pp costs) ;
+  let cpus = float @@ CPUSet.cardinal cpuset in
+  let nodes =
+    all_vms |> List.map (fun ((_, nodes), _) -> nodes) |> List.flatten
+  in
+  {costs with average= costs.average /. cpus; nodes}
+
+let cost_not_worse ~default c =
+  let worst = max default.worst c.worst in
+  let best = min default.best c.best in
+  let average = min default.average c.average in
+  D.debug "Default access times: %s; New plan: %s"
+    (Fmt.to_to_string pp default)
+    (Fmt.to_to_string pp c) ;
+  Alcotest.(
+    check int "The worst-case access time should not be changed from default"
```

Comment:
I'm a proponent of using `must` instead of `should` in tests. It's an error instead of a warning, after all.